Sources/NIOPosix/Bootstrap.swift

```diff
@@ -1352,6 +1352,482 @@ public final class ClientBootstrap: NIOClientTCPBootstrapProtocol {
+                            .flatMapThrowing {
+                                return try NIOAsyncChannel(
+                                    synchronouslyWrapping: channel,
+                                    backpressureStrategy: channelBackpressureStrategy,
+                                    isOutboundHalfClosureEnabled: isOutboundHalfClosureEnabled,
+                                    inboundType: channelInboundType,
+                                    outboundType: channelOutboundType
+                                )
+                            }
+                    }
+                }.flatMap { (asyncChannel: NIOAsyncChannel<ChildChannelInbound, ChildChannelOutbound>) in
+                    eventLoop.assertInEventLoop()
+                    return channel.registerAndDoSynchronously(body).map { asyncChannel }
+                }.flatMapError { error in
+                    channel.close0(error: error, mode: .all, promise: nil)
+                    return channel.eventLoop.makeFailedFuture(error)
+                }
+        }
+
+        if eventLoop.inEventLoop {
+            return setupChannel()
+        } else {
+            return eventLoop.flatSubmit {
+                setupChannel()
+            }
+        }
+    }
+}
+
+// MARK: AsyncChannel based connect with protocol negotiation
+extension ClientBootstrap {
+    /// Specify the `host` and `port` to connect to for the TCP `Channel` that will be established.
+    ///
+    /// - Parameters:
+    ///   - host: The host to connect to.
+    ///   - port: The port to connect to.
+    ///   - channelInitializer: A closure to initialize the channel which must return the handler that is used for negotiating
+    ///   the protocol.
+    /// - Returns: The protocol negotiation result.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @_spi(AsyncChannel)
+    public func connect<Handler: NIOProtocolNegotiationHandler>(
+        host: String,
+        port: Int,
+        channelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<Handler>
+    ) async throws -> Handler.NegotiationResult {
+        let loop = self.group.next()
+        let resolver = self.resolver ?? GetaddrinfoResolver(
+            loop: loop,
+            aiSocktype: .stream,
+            aiProtocol: .tcp
+        )
+
+        let connector = HappyEyeballsConnector<Handler.NegotiationResult>(
+            resolver: resolver,
+            loop: loop,
+            host: host,
+            port: port,
+            connectTimeout: self.connectTimeout
+        ) { eventLoop, protocolFamily in
+            return self.initializeAndRegisterNewChannel(
+                eventLoop: eventLoop,
+                protocolFamily: protocolFamily,
+                channelInitializer: channelInitializer
+            ) {
+                $0.eventLoop.makeSucceededFuture(())
+            }
+        }
+        return try await connector.resolveAndConnect().map { $0.1 }.get()
+    }
```

Comment:
Substantial amount of code is duplicated between this block and the block above. Can we find a way to define common functions that abstract the difference?